### DISCUSSION TITLE
Strange behavior when resource methods return null entities

### DISCUSSION THREAD

**Discussion by kalgon** [resteasy/resteasy#3024] [Q&A] [answered]
Hi, I have a question regarding the handling of null entities returned by resource methods.

Imagine the following resource which returns entities instead of `Response` objects:
```java
@Path("/foo")
public class MyResource {

    @GET
    @Path("/nonNull")
    public Integer getNonNull() {
        System.out.println("called non null");
        return 42;
    }

    @GET
    @Path("/null")
    public Integer getNull() {
        System.out.println("called null");
        return null;
    }

    @GET
    @Path("/void")
    public void getVoid() {
        System.out.println("called void");
    }
}
```
and this filter which changes the entity in the `ResponseContext`:
```java
public class MyFilter implements ContainerResponseFilter {

    @Override
    public void filter(ContainerRequestContext requestContext, ContainerResponseContext responseContext) {
        System.out.println("filtered");
        responseContext.setEntity("hello", new Annotation[0], MediaType.TEXT_PLAIN_TYPE); // will only be reflected if original entity is not null
        //responseContext.setEntity("hello"); // throws NPE when original entity is null
    }
}
```
If my `MyResource` and `MyFilter` are deployed as part of an `Application` (/api) and I call it like this:
```java
URI base = ...;
Client client = ClientBuilder.newClient(); // not autocloseable!!!
try {
    WebTarget target = client.target(base);
    System.out.println(target.path("/api/foo/nonNull").request().get(String.class)); // prints "hello"
    System.out.println(target.path("/api/foo/null").request().get(String.class)); // prints "null"
    System.out.println(target.path("/api/foo/void").request().get(String.class)); // prints "null"
} finally {
    client.close();
}
```
It seems that a returned null entity CANNOT be changed by the `ContainerResponseFilter`. Moreover, if a returned null entity is changed with `ResponseContext.setEntity(Object)` instead of `ResponseContext.setEntity(Object, Annotation[], MediaType)`, a `NullPointerException` is thrown.

Is this the expected behavior? I am testing with 4.7.5.Final.

**Comment by jamezp** [comment, marked as answer]
This definitely looks like a bug to me. I've filed for https://issues.redhat.com/browse/RESTEASY-3089 for this.

**Comment by jamezp** [comment]
@kalgon Coming back to this, I've discovered something. In your filter if you add:
```java
containerResponseContext.setStatusInfo(Response.Status.OK);
```

Then the entity should be set. I think the reason the entity stays `null` is because the status is 204 No Content. It seems reasonable to me that the filter is responsible for changing the status too if it adds an entity.
